src/conv/invokers/impl_gemm_dynamic.cpp

```diff
@@ -416,5 +416,298 @@ MakeImplGemmDynamicBackwardDataInvokerFactory<solver::TunableImplicitGemmGTCDyna
+    uint32_t shift_pack_0 =
+        magic_div_u32_pack_shift(mdiv_0.shift, mdiv_1.shift, mdiv_2.shift, mdiv_3.shift);
+
+    int dtile_iy = num_of_gemms > 1 ? static_cast<int>(mdiv_x_tilda.magic) : 0;
+    int dtile_ix = num_of_gemms > 1 ? static_cast<int>(mdiv_x_tilda.shift) : 0;
+    int dslice_y = num_of_gemms > 1 ? static_cast<int>(mdiv_y_tilda.magic) : y;
+    int dslice_x = num_of_gemms > 1 ? static_cast<int>(mdiv_y_tilda.shift) : x;
+    int dtile_h  = num_of_gemms > 1 ? static_cast<int>(mdiv_group_mn.magic) : h_tilda;
+    int dtile_w  = num_of_gemms > 1 ? static_cast<int>(mdiv_group_mn.shift) : w_tilda;
+
+    bool need_set_zero = false;
+    if(y < stride_h || x < stride_w || dilation_h != 1 || dilation_w != 1)
+        need_set_zero = true;
+    need_set_zero |= config.gemm_k_global_split > 0;
+
+    std::vector<OpKernelArg> opArgs;
+    opArgs.emplace_back(static_cast<void*>(nullptr));       // placeholder for input ptr
+    opArgs.emplace_back(static_cast<const void*>(nullptr)); // placeholder for weight ptr
+    opArgs.emplace_back(static_cast<const void*>(nullptr)); // placeholder for output ptr
+    opArgs.emplace_back(hi);
+    opArgs.emplace_back(wi);
+    opArgs.emplace_back(n);
+    opArgs.emplace_back(k / group);
+    opArgs.emplace_back(c / group);
+    opArgs.emplace_back(ho);
+    opArgs.emplace_back(wo);
+    opArgs.emplace_back(stride_h);
+    opArgs.emplace_back(stride_w);
+    opArgs.emplace_back(dilation_h);
+    opArgs.emplace_back(dilation_w);
+    opArgs.emplace_back(pad_h);
+    opArgs.emplace_back(pad_w);
+    opArgs.emplace_back(y);
+    opArgs.emplace_back(x);
+
+    opArgs.emplace_back(dtile_iy);
+    opArgs.emplace_back(dtile_ix);
+    opArgs.emplace_back(dilation_h / gcd_stride_dilation_h);
+    opArgs.emplace_back(dilation_w / gcd_stride_dilation_w);
+    opArgs.emplace_back(y_tilda);
+    opArgs.emplace_back(x_tilda);
+    opArgs.emplace_back(dtile_h);
+    opArgs.emplace_back(dtile_w);
+    opArgs.emplace_back(dslice_y);
+    opArgs.emplace_back(dslice_x);
+
+    opArgs.emplace_back(h_tilda_slice);
+    opArgs.emplace_back(w_tilda_slice);
+    opArgs.emplace_back(h_tilda_left);
+    opArgs.emplace_back(w_tilda_left);
+    opArgs.emplace_back(group);
+
+    opArgs.emplace_back(mdiv_0.magic);
+    opArgs.emplace_back(mdiv_1.magic);
+    opArgs.emplace_back(mdiv_2.magic);
+    opArgs.emplace_back(mdiv_3.magic);
+    opArgs.emplace_back(shift_pack_0);
+    opArgs.emplace_back(config.gemm_k_global_split);
+
+    return [opArgs, need_set_zero](const std::vector<Kernel>& kernels) mutable {
+        return [=](const Handle& handle, const AnyInvokeParams& primitive_parameters) mutable {
+            decltype(auto) data_ctx = primitive_parameters.CastTo<conv::DataInvokeParams>();
+            const auto& tensors     = data_ctx.tensors;
+            const auto ker          = handle.Run(kernels[0]);
+            float elapsed           = 0;
+
+#if MIOPEN_BACKEND_OPENCL
```

Comment:
Just curious: Does this switch (between OCL and HIP" have to be inside the invoker for each kernel, or it could be implemented in some kind of base class of invoker?

cc@atamazov

Reply:
This code/idea is borrowed from MLIR solver, https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/src/conv/invokers/mlir_impl_gemm.cpp#L156, where MLIR also need to modify the input/weight/output pointer inside the lambda invoker.

Reply:
@asroy 
> Does this switch (between OCL and HIP) have to be inside the invoker of each kernel

No, these details are hidden in the `Kernel` class.

Reply:
> This code/idea is borrowed from MLIR solver, https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/src/conv/invokers/mlir_impl_gemm.cpp#L156, where MLIR also need to modify the input/weight/output pointer inside the lambda invoker.

Generally, that code is a hack. I do not remember all the details, but the reason is that MLIR wants to modify instances of `MlirConvArgs`. You do not need this with `OpKernelArg` class.